File: 46-permutations/permutations.py
class Solution:

    def recurPermute(self, index: int, nums: List[int], ans: List[List[int]]):
        if index == len(nums):
            ans.append(nums[:])
            return 

        for i in range(index, len(nums)):
            nums[index], nums[i] = nums[i], nums[index]
            self.recurPermute(index + 1, nums, ans)
            nums[index], nums[i] = nums[i], nums[index]

    def permute(self, nums: List[int]) -> List[List[int]]:
        self.ans = []
        self.recurPermute(0, nums, self.ans)
        return self.ans
        

File: 46-permutations/permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {

        if (nums.length == 0) {
            List<List<Integer>> base = new ArrayList<>();
            base.add(new ArrayList<>());
            return base;
        }

        int first = nums[0];
        int[] rest = Arrays.copyOfRange(nums, 1, nums.length);
        List<List<Integer>> perms = permute(rest);

        List<List<Integer>> result = new ArrayList<>();

        for (List<Integer> perm : perms) {
            for (int i = 0; i <= perm.size(); i++) {
                List<Integer> temp = new ArrayList<>(perm);
                temp.add(i, first);
                result.add(temp);
            }
        }

        return result;

    }
}